src/hooks/useBreakpoint.ts

```diff
@@ -0,0 +1,42 @@
+import { useState, useCallback, useEffect } from 'react';
+import {
+  generateMediaQuery,
+  Mode as ModeType,
+  Breakpoint as BreakpointType
+} from '../components/Breakpoint/generateMediaQuery';
+
+type At = { at: string };
+type GreaterThan = { gt: string };
+type LessThan = { lt: string };
+
+export const discriminate = (
+  props: At | GreaterThan | LessThan
+): { mode: ModeType; breakpoint: BreakpointType } => {
+  if ('at' in props) return { mode: 'at' as ModeType, breakpoint: props.at as BreakpointType };
+  if ('lt' in props) return { mode: 'lt' as ModeType, breakpoint: props.lt as BreakpointType };
+  return { mode: 'gt' as ModeType, breakpoint: props.gt as BreakpointType };
+};
+
+export const useBreakpoint = (props: At | GreaterThan | LessThan) => {
+  const { mode, breakpoint } = discriminate(props);
+  const mediaQuery = generateMediaQuery({ mode, breakpoint });
+  const [matches, setMatches] = useState(true);
+
+  const handleChange = useCallback(
+    ({ matches }: MediaQueryListEvent): void => setMatches(matches),
+    []
+  );
+
+  useEffect(() => {
+    const mql = window.matchMedia(mediaQuery);
+    mql.addListener(handleChange);
+    setMatches(mql.matches);
+    return (): void => mql.removeListener(handleChange);
+  }, [handleChange, mediaQuery]);
+
+  return {
+    mode,
+    breakpoint,
+    matches
```

Comment:
I would return only `matches`. The other ones are irrelevant in almost all cases.
Like `useBreakpoint({lt: 'md'})` would return `{matches: true, mode: 'lt', breakpoint: 'md'}`. You see how `mode` and `breakpoint` are irrelevant, they are exactly what you pass in as params.
In the `Breakpoint` component you can use the `discriminate` function to get those values.